### PR TITLE
Add util contrib package for AppEngine standard.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Implements the first part of the issue: https://github.com/census-instrumentation/opencensus-java/issues/989

### CODE REVIEW

File: contrib/appengine-standard-util/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# OpenCensus AppEngine Standard Util
+[![Build Status][travis-image]][travis-url]
+[![Windows Build Status][appveyor-image]][appveyor-url]
+[![Maven Central][maven-image]][maven-url]
+
+The *OpenCensus AppEngine Standard Util for Java* is a collection of utilities for trace
+instrumentation when working with [AppEngine][appengine-url].
+
+## Quickstart
```
Comment (songy23) [inline contrib/appengine-standard-util/README.md]: Also specify the required JDK version here?

File: settings.gradle [hunk 2]
```diff
@@ -25,6 +26,8 @@ project(':opencensus-impl-lite').projectDir = "$rootDir/impl_lite" as File
 project(':opencensus-impl').projectDir = "$rootDir/impl" as File
 project(':opencensus-testing').projectDir = "$rootDir/testing" as File
 project(':opencensus-contrib-agent').projectDir = "$rootDir/contrib/agent" as File
+project(':opencensus-contrib-appengine-standard-util').projectDir =
```
Comment (songy23) [inline settings.gradle]: This artifact should be Java-8 only.

File: build.gradle [hunk 3]
```diff
@@ -74,6 +74,8 @@ subprojects {
         if (useErrorProne) {
             if (JavaVersion.current().isJava8Compatible()) {
                 it.options.compilerArgs += ["-XepDisableWarningsInGeneratedCode"]
+                // Exclude generated protobuf from error-prone checks.
+                it.options.compilerArgs += ["-XepExcludedPaths:.*/gen_gradle/.*"]
```
Comment (songy23) [inline build.gradle]: The generated protobuf should also be excluded from checker framework.

File: contrib/appengine-standard-util/src/main/proto/trace_id.proto [hunk 4]
```diff
@@ -0,0 +1,10 @@
+syntax = "proto3";
+
+option java_multiple_files = true;
+option java_package = "io.opencensus.contrib.appengine.standard.util";
+option java_outer_classname = "TraceProto";
+
+message TraceIdProto {
```
Comment (sebright) [inline contrib/appengine-standard-util/src/main/proto/trace_id.proto]: Why do we need to define a proto?  If CloudTraceContext uses a protocol buffer format for its trace ID, does it have a trace ID proto that we can reuse?

File: contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  private static final long INVALID_SPAN_ID = 0L;
+  private static final long INVALID_TRACE_MASK = 0L;
+
+  /**
+   * Converts AppEngine {@code CloudTraceContext} to {@code SpanContext}.
+   *
+   * @param cloudTraceContext the AppEngine {@code CloudTraceContext}.
+   * @return the converted {@code SpanContext}.
+   */
+  public static SpanContext toSpanContext(CloudTraceContext cloudTraceContext) {
```
Comment (sebright) [inline contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java]: Since this is already an OpenCensus library, I think it would be clearer to call the methods toCloudTraceContext and fromCloudTraceContext.  Those names are also more consistent with the method names in StatusConverter, i.e., to/fromGrpcCode and to/fromGrpcStatus.

File: contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+  /**
+   * Converts {@code SpanContext} to AppEngine {@code CloudTraceContext}.
+   *
+   * @param spanContext the {@code SpanContext}.
+   * @return the converted AppEngine {@code CloudTraceContext}.
+   */
+  public static CloudTraceContext toCloudTraceContext(SpanContext spanContext) {
+    if (spanContext == null) {
+      return new CloudTraceContext(INVALID_TRACE_ID, INVALID_SPAN_ID, INVALID_TRACE_MASK);
```
Comment (sebright) [inline contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java]: Same here

File: contrib/appengine-standard-util/src/test/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtilTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import com.google.apphosting.api.CloudTraceContext;
+import io.opencensus.trace.SpanContext;
+import io.opencensus.trace.SpanId;
+import io.opencensus.trace.TraceId;
+import io.opencensus.trace.TraceOptions;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link AppEngineCloudTraceContextUtilTest}. */
```
Comment (sebright) [inline contrib/appengine-standard-util/src/test/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtilTest.java]: s/AppEngineCloudTraceContextUtilTest/AppEngineCloudTraceContextUtil

File: contrib/appengine-standard-util/src/test/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+@RunWith(JUnit4.class)
+public class AppEngineCloudTraceContextUtilTest {
```
Comment (sebright) [inline contrib/appengine-standard-util/src/test/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtilTest.java]: It would help to also have some tests for converting null or invalid values, and converting a non-sampled context.

File: contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import io.opencensus.trace.TraceOptions;
+import java.nio.ByteBuffer;
+
+/**
+ * Utility class to convert between {@link io.opencensus.trace.SpanContext} and {@link
+ * CloudTraceContext}.
+ *
+ * @since 0.14
+ */
+public final class AppEngineCloudTraceContextUtil {
```
Comment (sebright) [inline contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java]: Most of the other public utility classes end with "Utils", such as ContextUtils and MonitoredResourceUtils.

File: contrib/appengine-standard-util/README.md (hunk 1 above)
Comment (bogdandrutu) [reply contrib/appengine-standard-util/README.md]: Now with java 1.7 I think no required to add this.

File: contrib/appengine-standard-util/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# OpenCensus AppEngine Standard Util
+[![Build Status][travis-image]][travis-url]
+[![Windows Build Status][appveyor-image]][appveyor-url]
+[![Maven Central][maven-image]][maven-url]
+
+The *OpenCensus AppEngine Standard Util for Java* is a collection of utilities for trace
+instrumentation when working with [AppEngine][appengine-url].
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply contrib/appengine-standard-util/README.md]: Next I would like to have a io.grpc.Context.Storage that also pushes the SpanContext to the appengine TLS (thread local storage) implementation.

Later we can add better stats support.

File: contrib/appengine-standard-util/build.gradle [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+[compileJava, compileTestJava].each() {
+    it.sourceCompatibility = 1.8
+    it.targetCompatibility = 1.8
+}
+
+dependencies {
+    compile project(':opencensus-api'),
+            libraries.appengine_api,
+            libraries.guava,
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply contrib/appengine-standard-util/build.gradle]: No. Thanks for catching.

File: contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java (hunk 9 above)
Comment (bogdandrutu) [reply contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java]: done.

File: contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java (hunk 5 above)
Comment (bogdandrutu) [reply contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java]: done.

File: contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+  /**
+   * Converts AppEngine {@code CloudTraceContext} to {@code SpanContext}.
+   *
+   * @param cloudTraceContext the AppEngine {@code CloudTraceContext}.
+   * @return the converted {@code SpanContext}.
+   */
+  public static SpanContext toSpanContext(CloudTraceContext cloudTraceContext) {
+    if (cloudTraceContext == null) {
+      return SpanContext.INVALID;
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java]: Do you want to checkNotNull?

File: contrib/appengine-standard-util/src/main/proto/trace_id.proto (hunk 4 above)
Comment (bogdandrutu) [reply contrib/appengine-standard-util/src/main/proto/trace_id.proto]: No public proto. See also https://github.com/GoogleCloudPlatform/cloud-trace-java/tree/master/services/appengine-standard-service/src/main/proto/cloud/trace/proto

File: contrib/appengine-standard-util/src/test/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtilTest.java (hunk 7 above)
Comment (bogdandrutu) [reply contrib/appengine-standard-util/src/test/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtilTest.java]: done.

File: contrib/appengine-standard-util/src/test/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtilTest.java (hunk 8 above)
Comment (bogdandrutu) [reply contrib/appengine-standard-util/src/test/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtilTest.java]: Done.

File: settings.gradle [hunk 13]
```diff
@@ -25,6 +26,8 @@ project(':opencensus-impl-lite').projectDir = "$rootDir/impl_lite" as File
...
+        "$rootDir/contrib/appengine-standard-util" as File
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply settings.gradle]: done.

File: contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java (hunk 6 above)
Comment (bogdandrutu) [reply contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java]: I will fix this when we agree on the previous comment.

File: contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (songy23) [reply contrib/appengine-standard-util/src/main/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtil.java]: Agree, checkNotNull() is consistent with other APIs.

File: contrib/appengine_standard_util/src/test/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtilsTest.java [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+              (byte) 0xa1,
+              (byte) 0xb5,
+              (byte) 0x04,
+              (byte) 0x65,
+              (byte) 0xcb,
+              (byte) 0x72
+            },
+            Long.MIN_VALUE,
+            // Trace enabled.
+            0L);
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply contrib/appengine_standard_util/src/test/java/io/opencensus/contrib/appengine/standard/util/AppEngineCloudTraceContextUtilsTest.java]: done.

**Comment by bogdandrutu** [comment]
PTAL. I think I fixed all comments.
